File: utils/Timer.py
import threading
from utils import Utils

timeInSeconds = 0
callback = None
timer = None


def set(ts, cb):
    global timeInSeconds, callback, timer
    if not Utils.isNumber(ts):
        print('Timer: time must be a number')
    if timer:
        print('Timer: Timer already set')
        return

    timeInSeconds = ts
    callback = cb
    timer = threading.Timer(0.0, countDown)


def add(ts):
    global timeInSeconds
    if not Utils.isNumber(ts):
        print('Timer: time must be a number')
    if not timer:
        print('Timer: No timer set')
        return
    timeInSeconds += ts


def substract(ts):
    global timeInSeconds
    if not Utils.isNumber(ts):
        print('Timer: time must be a number')
    if not timer:
        print('Timer: No timer set')
        return
    timeInSeconds -= ts


def start():
    if not timer:
        print('Timer: No timer set')
        return
    timer.start()


def stop():
    global timeInSeconds
    if not timer:
        print('Timer: No timer set')
        return
    timer.cancel()
    timeInSeconds = 0


def countDown():
    global timeInSeconds, callback, timer
    if timeInSeconds > 0:
        timeInSeconds -= 1
        mins, secs = divmod(timeInSeconds, 60)
        timeformat = '{:02d}:{:02d}'.format(mins, secs)
        print(timeformat, end='\r')
        timer = threading.Timer(1.0, countDown)
        timer.start()
    else:
        callback()


def reset():
    global timeInSeconds, callback, timer
    if timer:
        timer.cancel()
    timer = None
    timeInSeconds = 0
    callback = None

File: pi/utils/Timer.js
const TAG = 'Timer';
const utils = require('../utils/Utils');
const countdownIntervalMS = 1000;
let timerIntervalFunction = null;
let timeInSeconds = 0;
let isRunning = false;
let timerUpdateCallback = null;
let timerDoneCallback = null;

module.exports = class Timer {

    constructor(tdcb, tucb) {
        if(typeof tdcb != "function"){
            utils.log(TAG, "Timer done callback not set");
        }
        if(typeof tucb != "function"){
            utils.log(TAG, "Timer update callback not set");
        }
        timerUpdateCallback = tucb;
        timerDoneCallback = tdcb;
    }

    set (ts) {
        if(!Number.isInteger(ts)){
            utils.log(TAG, "Timer needs a number in seconds to be set");
            return;
        }
        if(isRunning){
            utils.log(TAG, "Timer is already running");
            return;
        }
        timeInSeconds = ts;
    }

    add (ts){
        if(!Number.isInteger(ts)){
            utils.log(TAG, "Timer needs a number in seconds to be set");
            return;
        }
        timeInSeconds += ts;
    }

    subtract (ts){
        if(!Number.isInteger(ts)){
            utils.log(TAG, "Timer needs a number in seconds to be set");
            return;
        }
        timeInSeconds -= ts;
    }

    start (){
        if(isRunning){
            utils.log(TAG, "Timer is already running");
            return;
        }
        isRunning = true;
        this.countdown();
        timerIntervalFunction = setInterval(this.countdown, countdownIntervalMS);
    }

    stop (){
        if(!isRunning){
            utils.log(TAG, "Timer is not running");
            return;
        }
        clearInterval(timerIntervalFunction);
        isRunning = false;
    }

    reset (){
        this.stop();
        timeInSeconds = 0;
    }

    getCurrentTime(){
        return timeInSeconds;
    }

    isRunning(){
        return isRunning;
    }

    countdown() {
        timeInSeconds--;
        utils.log(TAG, "Timer remaining: " + timeInSeconds);
        if(typeof timerUpdateCallback == "function"){
            timerUpdateCallback(timeInSeconds);
        }
        if(timeInSeconds <= 0){
            timeInSeconds = 0;
            clearInterval(timerIntervalFunction);
            isRunning = false;
            if(typeof timerDoneCallback == "function"){
                timerDoneCallback();
            }
        }
    }
};